services/auth/source/oauth2/store.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package oauth2
+
+import (
+	"fmt"
+	"net/http"
+
+	chiSession "gitea.com/go-chi/session"
+	"github.com/gorilla/sessions"
+)
+
+// SessionsStore creates a gothic store from our session
+type SessionsStore struct {
+}
+
+// Get should return a cached session.
+func (st *SessionsStore) Get(r *http.Request, name string) (*sessions.Session, error) {
+	chiStore := chiSession.GetSession(r)
+
+	rawData := chiStore.Get(name)
+	if rawData == nil {
+		return st.New(r, name)
+	}
+
+	oldSession, ok := rawData.(*sessions.Session)
+	if !ok {
+		return nil, fmt.Errorf("unexpected object in session: %v at name: %s", rawData, name)
+	}
+
+	// Copy over the old data into the session
+	session := sessions.NewSession(st, name)
+	session.ID = oldSession.ID
+	session.IsNew = oldSession.IsNew
+	session.Options = oldSession.Options
+	session.Values = oldSession.Values
+
+	return session, nil
+}
+
+// New should create and return a new session.
+//
+// Note that New should never return a nil session, even in the case of
+// an error if using the Registry infrastructure to cache the session.
+func (st *SessionsStore) New(r *http.Request, name string) (*sessions.Session, error) {
+	chiStore := chiSession.GetSession(r)
+
+	session := sessions.NewSession(st, name)
+	session.ID = chiStore.ID()
```

Comment:
Maybe we should only generate a new ID if there is no existing session.

Reply:
This is not generating a new ID.

Reply:
But what's the purpose of `session.ID = chiStore.ID()` here? If `rawData != nil` and `ok`, the `session` members including `ID` are all re-assigned.